### ISSUE TITLE
Feature Request: Context-sensitive escaping in docker-compose env files

### ISSUE THREAD

**Issue by bodograumann** [qetza/replacetokens#11]
We are using the replacetokens task to (among other things) replace placeholders in env files for docker compose.
These files have very specific requirements on escaping.
Currently we are using the "Replace Tokens" task, but it does not currently support escaping in these env files.
I realize this is a very specific use-case, but it is not possible to cover it with the custom-escaping config, due to the context-sensitive nature of the escaping.

Do you think it makes sense to add special escaping for `*.env` files here?

Basically what we need is the bash script below, which works well (up to a few edge cases). A bash script in the pipeline does not have access to all the variables though, afaik.

```bash
# Generate sed s-command to replace one literal with another one.
# The key is assumed to contain only characters [A-Z_]
function sub() {
  KEY="$1"
  VALUE=$(printf '%s' "$2" | sed -e 's/[\|&]/\\&/g')
  printf 's|#{%s}#|%s|g' "$KEY" "$VALUE"
}

# - Single quoted values need to have single quotes escaped.
# - Double quoted values need to have double quotes and backslashes escaped.
# - Double- und unquoted values need to have the dollar sign escaped.
VALUE_SQ=$(printf "%s" "$VALUE" | sed -e "s|'|\\\\'|g")
VALUE_DQ=$(printf "%s" "$VALUE" | sed -e 's|\\|\\\\|g' -e 's|"|\\"|g' -e 's|\$|$$|g')
VALUE_UQ=$(printf "%s" "$VALUE" | sed -e 's|\$|$$|g')
sed -i \
  -e '/^[^=]*='\''/ '"$(sub "$KEY" "$VALUE_SQ")" \
  -e '/^[^=]*="/ '"$(sub "$KEY" "$VALUE_DQ")" \
  -e "$(sub "$KEY" "$VALUE_UQ")" \
  "$FILE"
```

**Comment by qetza** [comment]
Hi @bodograumann,
The task is by default a simple find and replace without any context awareness to keep it simple as there would be too many contexts to support.

Are you using the task in Azure DevOps? If yes I don't see why your bash script will not have access to all variables as environment variables; the replacetokens task just uses the environment variables set by the agent.

I suppose it's not possible for you to "pre escape" the values in the variables?

**Comment by bodograumann** [comment]
Thanks for getting back to me. Yes, this is about Azure Devops :smile: 

Indeed pre-escaping will not work due to context-sensitivity. We would have to make assumptions about where the value will land in the end. Also in the future it would be forgotten, when new values will be added.

Turns out there are several reasons why the bash script did not have access to the variables.
First of all I assumed all keys where uppercase with underscores – they are not :rofl: :facepalm:.
Then there are also secrets. Only now I learned, that normal variables are mapped automatically to the env in scripts, but secret variables need to be mapped manually.

So when explicitely mapping the secrets to the script, it actually works.
I'll close this issue, assuming that the use-case is too narrow. If somebody else comes along with the same requirements, feel free to comment here and we can consider it again. (Or maybe I'll just share my bash solution then :smirk: )
